### PR TITLE
Add GitHub workflows for KMP

### PR DESCRIPTION
**Pull request by bitPogo**
## 📑 What does this PR do?

This addresses partially #12 by adding a setup which works for KMP.
It also adjusts the setups of KMP projects in order to make them properly work in the CI.

**This PR depends on #58**

# ✅ Checklist

- [x] My pull request adheres to the code style of this project
- [ ] My code requires changes to the documentation
- [ ] I have updated the documentation as required
- [ ] All the tests have passed

## 🧪 How can this PR been tested?

If the CI starts and runs consider it tested.

### CODE REVIEW

**Comment by Rafsanjani** [comment]
There are lots of changed files in this PR, making it quite difficult to see exactly what's changed.

**Review by Rafsanjani — APPROVED** [review]

**Review by Rafsanjani — APPROVED** [review]

File: .github/workflows/pull-requests.yml [hunk 1]
```diff
@@ -0,0 +1,49 @@
+name: PR Validation
+
+on:
+  push:
+    branches: [ "main", "development" ]
```
Comment (crocsandcoffee) [inline .github/workflows/pull-requests.yml:5]: just calling out we don't have a development branch yet.

**Review by crocsandcoffee — COMMENTED** [review]
Looks good overall - but I think we will want to fix the tests instead of commenting them all out. There's another PR that does that. Lets merge that and then reupdate this PR to ensure everything still works as expected

File: data/network/build.gradle.kts [hunk 2]
```diff
@@ -30,11 +30,12 @@ android {
     }
 }
 
+fun isIdea(): Boolean = System.getProperty("idea.fatal.error.notification") != null
```
[the remark this answers is not in the document]
Comment (bitPogo) [reply data/network/build.gradle.kts:33]: Well, consider that function as just there for now and gone with Electric Eel. If you want to still a comment over it, I can do that.

File: .github/workflows/pull-requests.yml (hunk 1 above)
Comment (bitPogo) [reply .github/workflows/pull-requests.yml:5]: Indeed - but I took that basically from the preexisting `continuous-delivery-pipeline.yml`. I thought it was there for a reason.

File: build.gradle.kts [hunk 3]
```diff
@@ -26,27 +26,3 @@ allprojects {
...
-// Kotlin compilation. You can add a source set to a target's compilation by connecting it
-// with the compilation's default source set using 'dependsOn'.
-// See https://kotlinlang.org/docs/reference/building-mpp-with-gradle.html#connecting-source-sets
-//
-// Remove log pollution until Android support in KMP improves.
-//
-subprojects {
-    afterEvaluate {
-        project.extensions.findByType<org.jetbrains.kotlin.gradle.dsl.KotlinMultiplatformExtension>()?.let { ext ->
-            ext.sourceSets.removeAll { sourceSet ->
```
[the remark this answers is not in the document]
Comment (bitPogo) [reply build.gradle.kts:42]: Yes and no - the error is not shown in Gradle anymore, however will be shown in AndroidStudio due to [this](https://issuetracker.google.com/u/1/issues/247996219), therefore is the `isIdea` method.

File: data/network/build.gradle.kts (hunk 2 above)
[the remark this answers is not in the document]
Comment (crocsandcoffee) [reply data/network/build.gradle.kts:33]: Got it - OK no worries. thanks

File: .github/workflows/pull-requests.yml (hunk 1 above)
Comment (crocsandcoffee) [reply .github/workflows/pull-requests.yml:5]: I think it was just a default template that was copied over - but doesn't hurt to just leave this is in for now.. all good 😄

File: data/network/src/androidMain/kotlin/social/androiddev/common/network/UrlStream.kt [hunk 4]
```diff
@@ -6,6 +6,6 @@ import io.ktor.client.engine.cio.CIO
 import io.ktor.client.request.get
 import java.io.ByteArrayInputStream
 
-actual suspend fun urlStream(url: String): ByteArrayInputStream = HttpClient(CIO).use {
-    ByteArrayInputStream(it.get(url).body())
+suspend fun String.toUrlStream(): ByteArrayInputStream = HttpClient(CIO).use {
```
Comment (thebino) [inline data/network/src/androidMain/kotlin/social/androiddev/common/network/UrlStream.kt:9]: why this change? Now it is not multiplatform anymore, right?

File: data/persistence/build.gradle.kts [hunk 5]
```diff
@@ -30,61 +30,67 @@ android {
     }
 }
 
+fun isIdea(): Boolean = System.getProperty("idea.fatal.error.notification") != null
```
Comment (thebino) [inline data/persistence/build.gradle.kts:33]: Do we need this function in every module? There should be a nicer way

File: .github/workflows/pull-requests.yml [hunk 6]
```diff
@@ -0,0 +1,49 @@
+name: PR Validation
```
Comment (thebino) [inline .github/workflows/pull-requests.yml:1]: Why not use the same filename so it is easier to track what has changed?
`.github/workflows/continuous-delivery-pipeline.yml`

**Review by thebino — CHANGES_REQUESTED** [review]
There are really a lof changes in this PR, would it be possible to break it down into several? I really would love to see the CI running, but there also a lot changes I would not approve

File: .github/workflows/pull-requests.yml (hunk 6 above)
Comment (bitPogo) [reply .github/workflows/pull-requests.yml:1]: Too generic, in my opinion. For example, I would put the release action in another flow. Maybe we can settle on `check` or `test`?

File: data/network/src/androidMain/kotlin/social/androiddev/common/network/UrlStream.kt (hunk 4 above)
Comment (bitPogo) [reply data/network/src/androidMain/kotlin/social/androiddev/common/network/UrlStream.kt:9]: It is now platform specific for Android/Jvm. Somewhere in `:ui` it is a shared source (since there are only Android/Jvm).

File: ui/common/src/androidMain/kotlin/social/androiddev/common/network/UrlStream.kt [hunk 7]
```diff
@@ -0,0 +1,5 @@
+package social.androiddev.common.network
+
+import java.io.ByteArrayInputStream
+
+actual suspend fun urlStreamOf(url: String) : ByteArrayInputStream = url.toUrlStream()
```
Comment (bitPogo) [inline ui/common/src/androidMain/kotlin/social/androiddev/common/network/UrlStream.kt:5]: @thebino, take a look here.

File: ui/common/src/desktopMain/kotlin/social/androiddev/common/network/UrlStream.kt [hunk 8]
```diff
@@ -0,0 +1,5 @@
+package social.androiddev.common.network
+
+import java.io.ByteArrayInputStream
+
+actual suspend fun urlStreamOf(url: String) : ByteArrayInputStream = url.toUrlStream()
```
Comment (bitPogo) [inline ui/common/src/desktopMain/kotlin/social/androiddev/common/network/UrlStream.kt:5]: @thebino, take a look here.

File: ui/common/src/commonMain/kotlin/social/androiddev/common/network/UrlStream.kt [hunk 9]
```diff
@@ -2,4 +2,4 @@ package social.androiddev.common.network
 
 import java.io.ByteArrayInputStream
 
-expect suspend fun urlStream(url: String) : ByteArrayInputStream
+expect suspend fun urlStreamOf(url: String) : ByteArrayInputStream
```
Comment (bitPogo) [inline ui/common/src/commonMain/kotlin/social/androiddev/common/network/UrlStream.kt:5]: @thebino, take a look here.

File: data/persistence/build.gradle.kts (hunk 5 above)
Comment (crocsandcoffee) [reply data/persistence/build.gradle.kts:33]: Maybe can solve this from the project level gradle

Comment (bitPogo) [reply data/persistence/build.gradle.kts:33]: On it already...but I bumped into some other issues....sooooo

Comment (bitPogo) [reply data/persistence/build.gradle.kts]: Now you take a look again

File: data/network/src/androidMain/kotlin/social/androiddev/common/network/UrlStream.kt (hunk 4 above)
Comment (thebino) [reply data/network/src/androidMain/kotlin/social/androiddev/common/network/UrlStream.kt:9]: but with the `expected/actual` pattern of KMP you can have different implementations for all platforms.
with your change it is just a function without taking care of KMP

Comment (bitPogo) [reply data/network/src/androidMain/kotlin/social/androiddev/common/network/UrlStream.kt:9]: This breaks ios in that layer, therefor it cannot be a shared source for all. But it can in `ui`.
